client/executor/wasmtime/src/imports.rs

```diff
@@ -57,19 +57,89 @@ pub(crate) fn resolve_imports(
 			)))
 		}
 
-		let resolved = match name {
-			"memory" => {
-				memory_import_index = Some(externs.len());
-				resolve_memory_import(store, &import_ty, heap_pages)?
+		if name == "memory" {
+			memory_import_index = Some(index);
+			externs.push((index, resolve_memory_import(store, &import_ty, heap_pages)?));
+			continue
+		}
+
+		match import_ty.ty() {
+			ExternType::Func(func_ty) => {
+				pending_func_imports.insert(name.to_owned(), (index, import_ty, func_ty));
 			},
 			_ =>
-				resolve_func_import(store, &import_ty, host_functions, allow_missing_func_imports)?,
+				return Err(WasmError::Other(format!(
+					"host doesn't provide any non function imports besides 'memory': {}:{}",
+					import_ty.module(),
+					name,
+				))),
 		};
-		externs.push(resolved);
 	}
+
+	let mut registry = Registry { store, externs, pending_func_imports };
+
+	H::register_static(&mut registry)?;
+	let mut externs = registry.externs;
+
+	if !registry.pending_func_imports.is_empty() {
+		if allow_missing_func_imports {
+			for (_, (index, import_ty, func_ty)) in registry.pending_func_imports {
+				externs.push((
+					index,
+					MissingHostFuncHandler::new(&import_ty)?.into_extern(store, &func_ty),
+				));
+			}
+		} else {
+			let mut names = Vec::new();
+			for (name, (_, import_ty, _)) in registry.pending_func_imports {
+				names.push(format!("'{}:{}'", import_ty.module(), name));
+			}
+			let names = names.join(", ");
+			return Err(WasmError::Other(format!(
+				"runtime requires function imports which are not present on the host: {}",
+				names
+			)))
+		}
+	}
+
+	externs.sort_unstable_by_key(|&(index, _)| index);
+	let externs = externs.into_iter().map(|(_, ext)| ext).collect();
+
 	Ok(Imports { memory_import_index, externs })
 }
 
+struct Registry<'a, 'b> {
+	store: &'a mut Store,
+	externs: Vec<(usize, Extern)>,
+	pending_func_imports: HashMap<String, (usize, ImportType<'b>, FuncType)>,
+}
+
+impl<'a, 'b> sp_wasm_interface::HostFunctionRegistry for Registry<'a, 'b> {
+	type State = StoreData;
+	type Error = WasmError;
+	type FunctionContext = HostContext<'a>;
+
+	fn with_function_context<R>(
+		caller: wasmtime::Caller<Self::State>,
+		callback: impl FnOnce(&mut dyn FunctionContext) -> R,
+	) -> R {
+		callback(&mut HostContext { caller })
+	}
+
+	fn register_static<Params, Results>(
+		&mut self,
+		fn_name: &str,
+		func: impl wasmtime::IntoFunc<Self::State, Params, Results>,
+	) -> Result<(), Self::Error> {
```

Comment:
Why do you return a result here? To be future proof?

Reply:
Hmm... good question! (:

Currently it can indeed only return an `Ok`; IIRC it's a vestigial remain from before one of the refactorings I did before I pushed this PR out where it could actually return an error.